### PR TITLE
Aws availability zone failure test (KNIP-622)

### PR DESCRIPTION
**Pull request by AviadP**

### CODE REVIEW

File: ocs_ci/utility/aws.py [hunk 1]
```diff
@@ -459,6 +459,22 @@ def get_ec2_instance_volumes(self, instance_id):
         volumes = instance.volumes.all()
         return [vol for vol in volumes]
 
+    def get_all_security_groups(self):
+        """
+        Get all security groups in AWS region
+
+        Returns:
+            all_sg_group: list: All security groups
```
Comment (ebenahar) [inline ocs_ci/utility/aws.py]: list: All security groups

File: tests/manage/cluster/nodes/test_az_failure.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        self.sanity_helpers = sanity_helpers.Sanity()
+
+    @pytest.fixture()
+    def teardown(self, request, ec2_instances, aws_obj):
+
+        def finalizer():
+            current_sg = aws_obj.store_security_groups_for_instances(self.instances_in_az)
+            if self.original_sgs != current_sg:
+                aws_obj.restore_instances_access(self.security_group_id, self.original_sgs)
+                logger.info(f"Access restores")
```
Comment (ebenahar) [inline tests/manage/cluster/nodes/test_az_failure.py]: f"Access to EC2 instances {self.instances_in_az} has been restored"

File: tests/manage/cluster/nodes/test_az_failure.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+            if self.security_group_id in aws_obj.get_all_security_groups():
+                logger.info(f"deleting: {self.security_group_id}")
```
Comment (ebenahar) [inline tests/manage/cluster/nodes/test_az_failure.py]: f"Deleting security group {self.security_group_id}"

File: tests/manage/cluster/nodes/test_az_failure.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+                aws_obj.delete_security_group(self.security_group_id)
+
+        request.addfinalizer(finalizer)
+
+    def test_availability_zone_failure(
+            self, aws_obj, ec2_instances, pvc_factory, pod_factory, teardown
```
Comment (ebenahar) [inline tests/manage/cluster/nodes/test_az_failure.py]: you have an extra indentation here

File: tests/manage/cluster/nodes/test_az_failure.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        # Storing current security groups for selected instances:
+        self.original_sgs = aws_obj.store_security_groups_for_instances(self.instances_in_az)
+        logger.info(f"Original security groups of selected instances: {self.original_sgs}")
+
+        # Blocking instances:
+        self.security_group_id = self.block_aws_availability_zone(aws_obj, self.instances_in_az)
+        logger.info("Access Blocked")
+
+        # Check cluster's health, need to be unhealthy at that point
+        assert self.check_cluster_health() == 0
```
Comment (ebenahar) [inline tests/manage/cluster/nodes/test_az_failure.py]:
```python
assert not self.check_cluster_health(), "Cluster is wrongly reported as healthy. EC2 Instances {self.instances_in_az} are blocked"
```

File: tests/manage/cluster/nodes/test_az_failure.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        self.instances_in_az = self.random_availability_zone_selector(aws_obj, ec2_instances)
+        logger.info(f"AZ selected, Instances: {self.instances_in_az} to be blocked")
+
+        # Storing current security groups for selected instances:
+        self.original_sgs = aws_obj.store_security_groups_for_instances(self.instances_in_az)
+        logger.info(f"Original security groups of selected instances: {self.original_sgs}")
+
+        # Blocking instances:
+        self.security_group_id = self.block_aws_availability_zone(aws_obj, self.instances_in_az)
+        logger.info("Access Blocked")
```
Comment (ebenahar) [inline tests/manage/cluster/nodes/test_az_failure.py]: f"Access to EC2 instances {self.instances_in_az} has been blocked"

File: tests/manage/cluster/nodes/test_az_failure.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        logger.info("Trying to delete resources on un-healthy cluster")
+        self.sanity_helpers.delete_resources()
+        logger.info("Resources Deleted")
+
+        # Restore access for blocked instances
+        aws_obj.restore_instances_access(self.security_group_id, self.original_sgs)
+        logger.info(f"Access restores")
+
+        # Check cluster's health, need to be healthy at that point
+        assert self.check_cluster_health() == 1
```
Comment (ebenahar) [inline tests/manage/cluster/nodes/test_az_failure.py]:
```python
assert self.check_cluster_health(), "Cluster is unhealthy"
```

File: tests/manage/cluster/nodes/test_az_failure.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+        security_group_id = aws_obj.create_security_group(group_name, dict_permissions, vpc_id)
+        aws_obj.block_instances_access(security_group_id, instances_in_az)
+
+        return security_group_id
+
+    def check_cluster_health(self):
+        try:
+            self.sanity_helpers.health_check()
+            return True
+        except Exception as e:
```
Comment (ebenahar) [inline tests/manage/cluster/nodes/test_az_failure.py]: This is very risky as you are catching any kind of exception

File: tests/manage/cluster/nodes/test_az_failure.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+        self.original_sgs = aws_obj.store_security_groups_for_instances(self.instances_in_az)
+        logger.info(f"Original security groups of selected instances: {self.original_sgs}")
+
+        # Blocking instances:
+        self.security_group_id = self.block_aws_availability_zone(aws_obj, self.instances_in_az)
+        logger.info(f"Access to EC2 instances {self.instances_in_az} has been blocked")
+
+        # Check cluster's health, need to be unhealthy at that point
+
+        assert not self.check_cluster_health(), \
```
Comment (ebenahar) [inline tests/manage/cluster/nodes/test_az_failure.py]: Please break the lines as follows:

```python
assert not self.check_cluster_health(), (
    "Cluster is wrongly reported as healthy. "
    "EC2 Instances {self.instances_in_az} are blocked"
)
```

File: tests/manage/cluster/nodes/test_az_failure.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+        security_group_id = aws_obj.create_security_group(group_name, dict_permissions, vpc_id)
+        aws_obj.block_instances_access(security_group_id, instances_in_az)
+
+        return security_group_id
+
+    def check_cluster_health(self):
+        try:
+            self.sanity_helpers.health_check()
+            return True
+        except CommandFailed as e:
```
Comment (ebenahar) [inline tests/manage/cluster/nodes/test_az_failure.py]: In order to make sure you catch the right failure, I suggest you look for a specific string in the exception. something like:

```python
except CommandFailed as e:
    if "Unable to connect to the server" in str(e):
        logger.warning("Cluster is not healthy")
        return False
```

File: tests/manage/cluster/nodes/test_az_failure.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+            print(e, "Cluster is not healthy")
+            return False
```
Comment (ebenahar) [inline tests/manage/cluster/nodes/test_az_failure.py]: Please use logging instead of print

File: ocs_ci/utility/aws.py [hunk 12]
```diff
@@ -459,6 +459,22 @@ def get_ec2_instance_volumes(self, instance_id):
         volumes = instance.volumes.all()
         return [vol for vol in volumes]
 
+    def get_all_security_groups(self):
+        """
+        Get all security groups in AWS region
+
+        Returns:
+            list: All security groups
```
Comment (MeridianExplorer) [inline ocs_ci/utility/aws.py]: Please add a newline before doc end

File: ocs_ci/utility/aws.py [hunk 13]
```diff
@@ -459,6 +459,22 @@ def get_ec2_instance_volumes(self, instance_id):
...
+        """
```
Comment (MeridianExplorer) [inline ocs_ci/utility/aws.py]: Please add a newline before the doc end

File: ocs_ci/utility/aws.py [hunk 14]
```diff
@@ -459,6 +459,22 @@ def get_ec2_instance_volumes(self, instance_id):
...
+        all_sg_group = list()
```
Comment (MeridianExplorer) [inline ocs_ci/utility/aws.py]: `security group group` sounds a bit redundant - maybe `sg_list`?

File: tests/manage/cluster/nodes/test_az_failure.py [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+from ocs_ci.framework.testlib import ManageTest, tier4, bugzilla
+from ocs_ci.ocs.exceptions import CommandFailed
+from tests import sanity_helpers
+
+logger = logging.getLogger(__name__)
+
+
+@tier4
+@pytest.mark.polarion_id("OCS-1287")
+@pytest.mark.skipif(
```
Comment (MeridianExplorer) [inline tests/manage/cluster/nodes/test_az_failure.py]: Please replace with the recently added `aws_platform_required` flag

**Comment by AviadP** [comment]
tested locally due to Jenkins BZ.
log attached
[az_pr_local.log](https://github.com/red-hat-storage/ocs-ci/files/3684885/az_pr_local.log)

**Pull request merged**